Add reactions in the /json/messages endpoint

kracekumar opened the pull request:
Since the [existing PR](https://github.com/zulip/zulip/pull/2511/files) went out of my hand due to conflicts and `git squash`. I am opening a new PR.

@timabbott @arpith Have a look.

smarx commented:
*Automated message from Dropbox CLA bot*

@kracekumar, thanks for the pull request! It looks like you haven't yet signed the Dropbox CLA. Please [sign it here](https://opensource.dropbox.com/cla/).

smarx commented:
*Automated message from Dropbox CLA bot*
				
@kracekumar, thanks for signing the CLA!

arpith commented:
@kracekumar do you need to fetch upstream and rebase?

kracekumar commented:
@arpith Thanks for the suggestion. The issue was missing `type hint` for the latest parameter in `build_message_dict`. Now all tests are [passing](https://travis-ci.org/kracekumar/zulip/builds/181864925).

arpith commented:
👍

arpith commented:
@timabbott can you take a look? (the old PR was https://github.com/zulip/zulip/pull/2511)

timabbott commented on zerver/tests/test_narrow.py:
```diff
@@ -355,6 +356,50 @@ def get_query_ids(self):
+                self.assertEquals(message['reactions'], 1)
+                self.assertEquals(message['reactions'][0]['emoji_name'],
+                                  reaction.emoji_name)
+                break
```
This loop is a problematic testing pattern, because if the message ID doesn't appear in the list, the test will silently pass.

kracekumar replied:
You're right! `If` condition was never matched. Debugging why this is failing.

```py
        stream = "Denmark"
        hamlet_email = "hamlet@zulip.com"
        othello_email = "othello@zulip.com"

        self.login(hamlet_email)
        message_id = self.send_message(hamlet_email, stream, Recipient.STREAM)
        self.login(othello_email)

        reaction = self.add_reaction_to_message(message_id,
                                                othello_email)
        self.login(hamlet_email)
        messages = self.get_and_check_messages(dict())
        message_to_assert = None
        for message in messages['messages']:
            if message['id'] == message_id:
                message_to_assert = message
                break

        self.assertEquals(message_to_assert['reactions'], 1)
        self.assertEquals(message_to_assert['reactions'][0]['emoji_name'],
                          reaction.emoji_name)
```

timabbott commented on zerver/tests/test_narrow.py:
```diff
@@ -355,6 +356,50 @@ def get_query_ids(self):
+        self.subscribe_to_stream(hamlet_email, stream)
+
+        self.login(othello_email)
+        self.subscribe_to_stream(othello_email, stream)
```
If you use one of the default streams like `Denmark`, you shouldn't need to subscribe the users to it, and thus can save half this code.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -586,6 +590,82 @@ def test_applying_markdown(self):
+        sending_client = make_client(name="test suite")
+
+        needed_ids = []
+        for i in range(100):
```
This is a pretty expensive test (doing 200 database queries); can we do a much smaller number (e.g. 5) and still get the same testing benefit?

timabbott commented on zerver/models.py:
```diff
@@ -1019,6 +1046,14 @@ class Reaction(ModelReprMixin, models.Model):
+        fields = ['message_id', 'emoji_name', 'user_profile__email',
+                  'user_profile__id', 'user_profile__full_name']
+        return Reaction.objects.filter(message_id__in=needed_ids).values(
+            *fields)
```
I would probably not line-wrap this piece.

timabbott commented on zerver/lib/message.py:
```diff
@@ -78,6 +78,7 @@ def to_dict_uncached_helper(message, apply_markdown):
                 recipient_id = message.recipient.id,
                 recipient_type = message.recipient.type,
                 recipient_type_id = message.recipient.type_id,
+                reactions = message.reaction_set.all()
```
what is `.reaction_set`?

kracekumar replied:
`reaction_set` is wrong method here, I am replacing it with `Reaction.get_raw_db_rows([message.id]`

kracekumar commented:
@timabbott I have added the call to `update_to_dict_cache`. The integration test is modified now(different logic) and fixed the problematic loop. Happy to hear feedback and act on.

timabbott commented on zerver/tests/test_narrow.py:
```diff
@@ -388,8 +389,8 @@ def get_and_check_messages(self, modified_params):
             for field in ("content", "content_type", "display_recipient",
-                          "avatar_url", "recipient_id", "sender_full_name",
-                          "sender_short_name", "timestamp"):
+                         "avatar_url", "recipient_id", "sender_full_name",
```
This whitespace is a bit off here.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1254,7 +1263,6 @@ def check_message(sender, client, message_type_name, message_to,
         #     return json_error(_("Invalid subject name"))
 
         stream = get_stream(stream_name, realm)
-
```
this file has a few unnecessary whitespace changes in it.
